### PR TITLE
[JN-634] handling kit bad address errors in the UX

### PR DESCRIPTION
**Pull request by devonbush**
#### DESCRIPTION

This upgrades the error handling of our Pepper API connections, and fixes some stub and test methods that were out of date. Notably:

1. we now have separate Exception classes for "Pepper returned an error" vs. "Pepper returned success, but we don't understand what it said".   
2. We now use ControllerAdvice to handle mapping the PepperApiExceptions to a suitable response.  

The biggest struggle was wrestling with how our openapi source generation clashed with how Spring wants us to handle exceptions.  Despite an hour of research and trying, I was unable to get the generators to add "throws Exception" to any controller method (docs say you should just be able to add "unhandledException" as a config property -- this did not work). This meant that PepperApiException had to be a RuntimeException, otherwise the controller method would have had to catch it and rethrow it as a runtime exception or build an response from scratch, which seemed cumbersome and/or defeating the purpose of configurable exception handling.  I'd love to find a better solution, but this is better than what we had before, and we need to move this along.

Another area that could use improvement in the future is differentiating our response code based on whether Pepper gave us a 400 or 500 error.

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

1.  redploy ApiAdminApp
2. repopulate ourhealth
3. go to `https://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants/OHBADR`
4. click "request kit" and request a kit
5. confirm a message appears about the address being invalid
image

### CODE REVIEW

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:74]: Not for this PR, but this becomes a (potentially) interesting error/exception handling case. Presuming this will be caught (at the latest) as part of response processing (there are other cases, like cron jobs and message handling, but let's put those aside for now), we will want to be consistent on response handling. IOW, is this a 400 or 500? 
If we move to more specific Juniper specific exception classes, we do not lose any expressiveness (the message conveys what is needed), but we are able to specify whether this is an internal error or bad user input. 

In this particular case it is unclear. Should the admin have not asked for a kit for an enrollee without a profile, or is this more like an assert: enrollees should have profiles at this point.

Does that make sense?

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:86]: Does it make sense to catch `PepperApiException` here if there is something sensible we can do with it? Seems like we'd want to at least handle 400 and 500 responses differently, and further up the stack is probably not the place to for that.
By re-throwing different exceptions here we can also convert the message to something that makes sense to the user,  and (implicitly) dictate the response handling.

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/LivePepperDSMClient.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/pepper/LivePepperDSMClient.java]: Is this really a `PepperApiException`, since it seems to be a fundamental contract misunderstanding? IOW, if this lands back with the user, because of how we handle `PepperApiException` further up the stack, then what are they supposed to do to fix this?

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java:7]: I think the error handling gets easier if we use this only at the lowest levels of the Pepper interface code (throw and catches), and convert the errors to something more generic that the rest of the stack can confidently process. Not wed to the concept, but I suggested having two uber exception classes like `JuniperBadRequest` and `JuniperInternalError` (or whatever naming works).  
This requires that the higher levels have no need to process a Pepper error any differently from the error from any service, but I suspect that is true for most/all service errors. (And in cases where it is not true, we might want to take a closer look at the cases.

**Review by denniscunningham — COMMENTED** [review]
I am not done with the review, but you asked for early feedback...

Following up on our conversation this afternoon, I think this gets less complicated if we move to more generic exception classes geared towards response handling. (See my comments for more details.) This is easy to do gradually, so it would not be a major change to introduce that new technique in this PR.

It is entirely possible I missed an important detail, so happy to learn more.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:74]: Agree -- also because this check likely will get finer-scale -- sure they have a profile, but what if they have a profile with no address?  or an address with no street?

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:86]: Hmm... I'm not sure what we'd do with it.  Especially thinking that this is a service that will later on be invoked by automatic processes, I think we want to leave it up to the caller how to handle the various flavors of PepperApiException.

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/LivePepperDSMClient.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/pepper/LivePepperDSMClient.java]: Yeah, I think you're right that this is a PepperParseException

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java:7]: Interesting -- maybe we have those two base classes and then PepperApiException and friends extend them?  Then we don't have to do conversion -- each layer of the stack can operate on whatever granularity it likes?

File: core/src/main/java/bio/terra/pearl/core/dao/DaoUtils.java
Comment (MatthewBemis) [inline core/src/main/java/bio/terra/pearl/core/dao/DaoUtils.java:23]: What's the benefit of going to the database to get a UUID instead of using `java.util.UUID.randomUUID()`? Is it to ensure consistency in how/which version of UUID is generated?

**Review by MatthewBemis — APPROVED** [review]
One question, otherwise looks solid.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:86]: But what is a different flavor? Just in terms of encapsulation, why would code external to this service need to understand how to interpret responses from Pepper, particularly if they arrive in the same packaging, PepperApiException.

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java
Comment (denniscunningham) [reply core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java:7]: Sure, if that works for you. I suspect that additional shell is not necessary, but there is not a lot of downside to doing that way (unless the Spring exception handling stuff cannot handle the inheritance properly).

File: core/src/main/java/bio/terra/pearl/core/dao/DaoUtils.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/dao/DaoUtils.java:23]: yeah, I think UUID.randomUUID would work just as well, this is just being super careful.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:86]: Ah, maybe the code/comments aren't clear about the control flow -- we need to catch the PepperParseException because that means that the request to Pepper was successful.  And if the request was successful, we need to make sure we create the matching record in our database for bookkeeping purposes.    And likewise, we need to return a 20x status code since the operation overall was a success, so rethrowing something here just means we'd have to catch it again at the controller.  Telling an admin "your kit request was successful but DSM included some unexpected fields in its response" isn't particularly useful--I think logging as error so it gets picked up in alerting/triage is probably best because chances are the error is going to be very inside-baseball.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:74]: As things stand now this exception will whiz past the exception handling in `KitExtService.requestKits`, and I assume respond with a 500. Is that what we want in the multi-enrollee case? I can't tell if this is an assert or a viable case.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:86]: Makes sense. If we get a muddled response do we have enough info to make a useful record in the DB?
Do we have alerting setup for 500 responses and logging at levels >= error?

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:89]: You probably want to include the exception in the error logging here and below?

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:174]: This probably deserves a comment, similar to the above, since swallowing exceptions should be considered exceptional (ahem).

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:249]: This is a great place for `JuniperInternalError`, when we invent it.

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java:24]: Are you using these methods now (where you include the HttpStatus)? I think it's worth discussion on alternative approaches.

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperErrorResponse.java
Comment (denniscunningham) [inline core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperErrorResponse.java]: Not sure I understand why this extends `PepperReponse`, since it appears only be used as a field in `PepperErrorResponse`.

**Review by denniscunningham — COMMENTED** [review]
A few more questions. I realize we want to get this in soon, so I'm trying to not hold this up and move on to an exception handling discussion later. This is a clear improvement over the existing code, which I assume is our main objective.

I did not manually test a multi-enrollee kit request, so let me know if you would like me to do that.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:74]: It's an IllegalArgumentException, which AbstractGlobalExceptionHandler will turn into a BAD_REQUEST response.  You're right that we probably want to catch this in the Ext service for multi-requests.  Will add.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:86]: we don't yet have alerting -- just the weekly log review.  We'll likely want to up that to daily and add alerting once OurHealth traffic ramps up.

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:89]: we're very leery of including exceptions and responses in logs--the directive from AppSec/Compliance is to avoid logging PII/PHI, and logging the exception/response here seems like a reasonable risk that the response from Pepper would have PII/PHI in it

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperApiException.java:24]: yes, capturing the httpStatus that pepper returned seems very useful, in terms of differentiating failures due to auth, input, or other.  Happy to talk about other ways to pass that info along

File: core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperErrorResponse.java
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/pepper/PepperErrorResponse.java]: I can't remember why I did that either, removing

**Review by devonbush — COMMENTED** [review]
thanks for the good comments, just pushed some cleanups

File: core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java
Comment (denniscunningham) [reply core/src/main/java/bio/terra/pearl/core/service/kit/KitRequestService.java:89]: Not sure I agree with that, but is the stack trace helpful here? If so, we could just add that part. (Perhaps not for this PR.)

File: api-admin/src/main/java/bio/terra/pearl/api/admin/service/kit/KitExtService.java
Comment (denniscunningham) [inline api-admin/src/main/java/bio/terra/pearl/api/admin/service/kit/KitExtService.java:50]: You could keep the `PepperApiException` processing and add `Exception` processing since that seems useful?

**Review by denniscunningham — APPROVED** [review]
Looks good.
Just a couple comments. Approving so you can decide if/when to address those.

File: api-admin/src/main/java/bio/terra/pearl/api/admin/service/kit/KitExtService.java
Comment (devonbush) [reply api-admin/src/main/java/bio/terra/pearl/api/admin/service/kit/KitExtService.java:50]: good point.  we'll see how this shakes out once we have a better overall exception-handling strategy.

**Pull request merged**
